[KEP-2400]: Swap-aware memory eviction (with no additional APIs)

iholder101 opened the pull request:
#### What type of PR is this?
/kind feature
/sig node

#### Problem and background
Prior to this PR, kubelet's eviction manager completely overlooked swap memory, leading to several issues:
* **Inaccessible Swap**: The memory eviction threshold is configured in such a way that swap is never triggered during node-level pressure, as eviction occurs before the node starts swapping memory.
* **Unfairness & Instability**: The eviction manager may evict the "wrong" or innocent pods, failing to address the actual memory pressure.
* **Unexpected Behavior**: Pods that exceed their memory limits (with regular and swap memory) are not evicted first, even though they would immediately get killed if swap were not used.

In addition, this PR serves as an alternative to https://github.com/kubernetes/kubernetes/pull/128137. In that PR, swap is presented as a standalone signal alongside memory. However, I believe this approach is problematic, as it attempts to completely separate memory and swap memory. In reality, the two are inherently connected and should be addressed as a single issue. As a small example for how the two are inherently connected, swap is never used until the memory is full.

#### What this PR does / why we need it:

This PR addresses the above issues by making the eviction manager swap-aware. The proposed logic is fully backward compatible and requires no additional configuration, making it completely transparent to the user.

#### The main idea
Let `accessible swap` be the amount of swap that is accessible by pods according to the LimitedSwap swap behavior [[1](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#steps-to-calculate-swap-limit)]. Note that the amount of accessible swap changes in time according to the pods running on the node.

**Triggering evictions**: The eviction manager considers `accessible swap` as additional memory capacity.
**Eviction ordering**: The [eviction order](https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#pod-selection-for-kubelet-eviction) will be defined as follows (for more info, see discussion [here](https://github.com/kubernetes/enhancements/pull/4701#discussion_r1944859336)):
```
The kubelet uses the following parameters to determine the pod eviction order:
1. Whether the pod's resource usage (memory usage + swap usage) exceeds requests (memory requests + swap requests)
2. [Pod Priority](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/)
3. The pod's resource usage (memory usage + swap usage) relative to requests (memory requests + swap requests)
```


<!--
Older version:

- Eviction manager considers `accessible swap` as additional memory capacity.
- In the context of ranking pods for evictions, swap memory is considered as additional "regular" memory. This is relevant for checking whether memory requests are exceeded [[2](https://github.com/kubernetes/kubernetes/blob/d8093cc40394b8e25a864576fe6a38306730d3cb/pkg/kubelet/eviction/helpers.go#L684)] or for identifying which pods uses more memory [[3](https://github.com/kubernetes/kubernetes/blob/d8093cc40394b8e25a864576fe6a38306730d3cb/pkg/kubelet/eviction/helpers.go#L703)].
- When the memory eviction threshold is reached, pods that use more memory and swap than their memory limits are being killed first. Note that while there is no memory pressure using more than the pods limits is perfectly ok.
-->
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #120800

#### Special notes for your reviewer:
- I recommend reviewing this PR per commit.
- ~Please don't mind the last commit. It cherry picks the fix from https://github.com/kubernetes/kubernetes/pull/129486 and will be removed after this PR is merged.~ the PR is now merged and rebase on top of this one.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Make eviction mechanism swap-aware
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md
```

iholder101 committed:
Treat accessible swap as memory capacity

Accessible swap is the amount of swap that is
accessible by Kubernetes pods, according to the
LimitedSwap swap behavior.

This PR treats accessible swap as an additional node
memory capacity in the context of the eviction manager
making memory signal observesations.

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
Delete stress pod immediately as best-effort

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
Reset swap before each test

Turn swap on and off and sleep for a few seconds.
This helps stabalize the tests which will start
with a fresh unused swap space.

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
Eviction tests: small refactor

This small refactor:
- Adds swap log statistics.
- Adds a post pressure validation function.
- Adds a pre pods modification function.

The later can be used in order to perform
validation on the given pod right after a
pressure condition is met, or to perform
changes to pods before creation.

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
Add e2e tests

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
/test pull-kubernetes-node-swap-conformance-fedora-serial
/test pull-kubernetes-node-swap-conformance-ubuntu-serial

iholder101 commented:
/test pull-kubernetes-node-swap-conformance-fedora-serial
/test pull-kubernetes-node-swap-conformance-ubuntu-serial

iholder101 commented:
/test pull-kubernetes-node-swap-conformance-fedora-serial
/test pull-kubernetes-node-swap-conformance-ubuntu-serial

pacoxu commented:
/cc @harche @kannon92

pacoxu commented:
> In addition, this PR serves as an alternative to #128137. In that PR, swap is presented as a standalone signal alongside memory.

Should there be different eviction strategies? 
For instance,  a pod A with 2Gi memory and 2Gi swap limit, and another pod B is with 2Gi memory, 2 Gi swap limit.
At a time, pod A is using 1.9Gi memory with only 1Gi swap, pod B is using 1Gi memory and 1.9Gi swap. 
Which should be evicted first? Should this depends on swap usage and memory usage, or should this be configurable? Do you know the best practice?

iholder101 commented:
Hey @pacoxu! Thanks for your review!

Firstly, I want to highlight that we had initially agreed to defer handling swap evictions to a follow-up KEP. However, last-minute concerns ultimately blocked the swap GA. In my opinion, we should aim for a minimal swap-aware eviction logic that would suffice for GA and then incrementally extend it over time, rather than trying to create a mechanism that addresses all possible cases from the start. This approach is similar to how we currently have a single `LimitedSwap` behavior that can (and probably should!) be extended in the future.

> > In addition, this PR serves as an alternative to #128137. In that PR, swap is presented as a standalone signal alongside memory.
> 
> **EDITTED**
> 
> Should Kubernetes support different eviction strategies for handling swap memory usage?

It could. These alternatives come in to mind:
1. We can tie up the eviction logic to the swap behavior. In other words, we can say that the eviction strategy presented in this PR will be active when `LimitedSwap` is set. This way we can define more swap behaviors and corresponding eviction strategies in the future.
2. We can add an `evictionStrategy` field to Kubelet's [swap config](https://github.com/kubernetes/kubernetes/blob/f3cbd79db7f0c86a2d3602fdff6b174543d2cf1c/pkg/kubelet/apis/config/types.go#L685). Currently the only strategy that would be supported (and that would be the default) is the one presented in this PR.
3. Avoid adding an eviction strategy specifically for swap and instead introduce a plugin system for the eviction manager in a follow-up KEP. This will open the door for interesting ideas like IO eviction and custom user-made eviction strategies. I am more than willing to drive this KEP after this one is GAed.

> For example:
> 
> 1. Pod A has a memory limit of 2Gi and a swap limit of 2Gi. At a given time, it is using 1.9Gi of memory and 1Gi of swap.
> 2. Pod B has the same limits (2Gi memory and 2Gi swap) but is using 1Gi of memory and 1.9Gi of swap.
> 
> In this scenario:
> 
> 1. Which Pod should be evicted first? Should the decision prioritize Pods that use more swap memory (like Pod B) or those that use more regular memory (like Pod A)?
> 2. Should the eviction logic consider swap usage and memory usage differently? For instance, should swap usage have a lower weight than regular memory usage, or should they be treated equally?

In this PR's implementation both pods exceed their memory limits, so they'll be ranked the highest for eviction.

In essence, prioritizing such decisions is challenging without more context, which we often lack. Consider these two scenarios:
1. Pod A is actively using & dirtying its memory. Pod B doesn't do much work. Therefore, when there's a node-level pressure, the kernel decides to swap more of B's memory. In this case we probably want to prefer evicting pod A.
2. Pod A is not doing much work. Pod B is continuously crossing its memory limits which lead to the memory being swapped. Due to node pressure, the kernel decides to swap some of A's memory. In this case we probably want to prefer evicting pod B.

The key point here is that a pod's memory can be swapped either because it's a busy pod contributing to node pressure or because it's an idle pod that's completely innocent. Creating weights between swapped memory and "regular" memory is very hard because of that, and I'm not sure there's an easy way of configuring it.

> 3. Is it necessary to make this behavior configurable? For example, allowing users to define weights for swap versus regular memory usage in eviction decisions.

I'm not convinced it is, but happy to be convinced otherwise.
I also think we shouldn't rush forward and stick to the [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) principle as a first step.

> 4. What are the best practices for handling swap memory in Kubernetes eviction strategies? Are there specific guidelines or recommendations for balancing swap and regular memory usage in resource-constrained environments?

In further to the above, most applications are not aware to whether they use swap or not. Swap is a resource that is transparently managed by the kernel, which is very good and experience doing that job. So in Kubernetes, just like in any other system, swap needs to be transparent to the workloads that don't need to concern themselves with such decisions.

haircommander commented on pkg/api/v1/resource/helpers.go:
```diff
@@ -26,8 +26,14 @@ import (
+type resourceType string
+
+const (
+	resourceTypeReuqest resourceType = "reuqest"
```
request* in both the variable name and value

iholder101 replied:
Done!

haircommander commented on test/e2e_node/swap_test.go:
```diff
@@ -65,6 +69,33 @@ var _ = SIGDescribe("Swap", "[LinuxOnly]", nodefeature.Swap, framework.WithSeria
+
+		framework.Logf("Swap is reset. Used swap bytes (should be near 0): %d", usedSwapBytes)
+
+		secondsToSleep := 10
```
without thinking too deeply about it,10 seems pretty high for this. I assume you workshopped it and that's as low as it works reliably?

iholder101 replied:
After testing in https://github.com/kubernetes/kubernetes/pull/129634, reduced it to 3 seconds.

iholder101 commented on pkg/kubelet/eviction/helpers.go:
```diff
@@ -671,6 +680,10 @@ func (ms *multiSorter) Less(i, j int) bool {
 	priority2 := corev1helpers.PodPriority(p2)
+
+	klog.V(3).InfoS("eviction ranking by priority", "pod", p1.Name, "namespace", p1.Namespace, "priority", priority1)
+	klog.V(3).InfoS("eviction ranking by priority", "pod", p2.Name, "namespace", p2.Namespace, "priority", priority2)
```
Raised to 5. Is it reasonable?

iholder101 commented:
> We probably should think about how to guard this code with a feature gate.

Great point!
Added :+1:

iholder101 commented:
/test pull-kubernetes-node-swap-conformance-fedora-serial
/test pull-kubernetes-node-swap-conformance-ubuntu-serial

iholder101 commented:
Rebased.

/test pull-kubernetes-node-swap-conformance-fedora-serial
/test pull-kubernetes-node-swap-conformance-ubuntu-serial

iholder101 commented:
/test pull-kubernetes-node-swap-conformance-fedora-serial
/test pull-kubernetes-node-swap-conformance-ubuntu-serial

iholder101 commented:
/test pull-kubernetes-node-swap-conformance-fedora-serial
/test pull-kubernetes-node-swap-conformance-ubuntu-serial

kannon92 commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -238,6 +239,40 @@ func (m *managerImpl) IsUnderPIDPressure() bool {
+		accessibleSwap += podAccessibleSwap
+	}
+
+	klog.InfoS("Node's accessible swap", "number of pods", len(pods), "accessible swap", accessibleSwap)
```
I don't love V(3) in this function. Should we make this higher?

Having lots of logs could remove important events from logs since they would get rotated in a journal.

kannon92 replied:
V(5) is probably good.

iholder101 replied:
Done :)

kannon92 reviewed (commented):
Can we had a presubmit that will verify that memory evictions tests will work on a swap enabled node?

I'd like to confirm that evictions due to Guaranteed/BestEffort pods would still occur if there is memory pressure. I think you could reuse the existing eviction tests and just add a new job (or use existing eviction presubmit to test memory evictions).

iholder101 commented:
> Can we had a presubmit that will verify that memory evictions tests will work on a swap enabled node?
> 
> I'd like to confirm that evictions due to Guaranteed/BestEffort pods would still occur if there is memory pressure. I think you could reuse the existing eviction tests and just add a new job (or use existing eviction presubmit to test memory evictions).

We have the swap-serial lanes (`pull-kubernetes-node-swap-fedora-serial` and `pull-kubernetes-node-swap-ubuntu-serial`), but it seems that they skip eviction tests [1][2].

Perhaps it's a good idea to remove this skip and ensure that eviction tests are running on these lanes.

Note: these lanes are not the swap-conformance lanes, they run "regular" tests but on swap-enabled nodes.

-----

[1] https://testgrid.k8s.io/sig-node-cri-o#kubelet-gce-e2e-swap-fedora-serial
[2] See `--skip="\[Flaky\]|\[Slow\]|\[Benchmark\]|\[NodeFeature:Eviction\]` in [build log](https://storage.googleapis.com/kubernetes-ci-logs/logs/ci-kubernetes-node-swap-fedora-serial/1879478157334024192/build-log.txt)

iholder101 commented:
/test pull-kubernetes-node-swap-conformance-fedora-serial
/test pull-kubernetes-node-swap-conformance-ubuntu-serial

iholder101 commented:
Rebased with the fix https://github.com/kubernetes/kubernetes/pull/129486 after the PR had merged.

/test pull-kubernetes-node-swap-conformance-fedora-serial
/test pull-kubernetes-node-swap-conformance-ubuntu-serial

ffromani commented:
/triage accepted
/priority important-longterm

rphillips commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -302,8 +337,24 @@ func (m *managerImpl) synchronize(diskInfoProvider DiskInfoProvider, podFunc Act
+			nodeTotalSwap := *summary.Node.Swap.SwapAvailableBytes + *summary.Node.Swap.SwapUsageBytes
+			accessibleSwap, err = calcAccessibleSwap(activePods, nodeTotalMemory, nodeTotalSwap)
+			if err != nil {
+				klog.ErrorS(err, "Eviction manager: failed to calculate accessible swap")
```
err isn't returned here and there might be a side-effect of implicitly setting accessibleSwap to zero. I suggest we explicitly set accessibleSwap on error here and return (or comment) that the code falls through.

iholder101 replied:
Done!

rphillips commented on pkg/kubelet/eviction/eviction_manager.go:
```diff
@@ -302,8 +337,24 @@ func (m *managerImpl) synchronize(diskInfoProvider DiskInfoProvider, podFunc Act
+				klog.ErrorS(err, "Eviction manager: failed to calculate accessible swap")
+			}
+		} else {
+			klog.InfoS("Summary stats do not contain node and swap capacity, skipping accessible swap calculation and assuming no swap available")
```
Perhaps log level (2) for lines 351 and 353?

rphillips commented:
Overall, lgtm.

iholder101 committed:
Introduce SwapCalculator interface

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
Refactor: Use SwapCalculator interface in kuberuntime_container_linux.go

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
Add swap to memory usage, accessible swap to memory request

This PR treats used swap memory as an additional
memory usage in the context of the eviction manager
ranking pods for eviction.

In addition, a pod's accessible swap is considered
as an additional memory request.

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
/test pull-kubernetes-node-swap-conformance-fedora-serial
/test pull-kubernetes-node-swap-conformance-ubuntu-serial

iholder101 commented:
/test pull-kubernetes-node-swap-conformance-fedora-serial
/test pull-kubernetes-node-swap-conformance-ubuntu-serial
/test pull-kubernetes-unit
/test pull-kubernetes-typecheck
/test pull-kubernetes-verify
/test pull-kubernetes-node-e2e-containerd

iholder101 commented:
As per the discussions that took part in the KEP update (see discussion [here](https://github.com/kubernetes/enhancements/pull/4701#discussion_r1944859336)), the [eviction order](https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#pod-selection-for-kubelet-eviction) will be defined as follows:
```
The kubelet uses the following parameters to determine the pod eviction order:
1. Whether the pod's resource usage (memory usage + swap usage) exceeds requests (memory requests + swap requests)
2. [Pod Priority](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/)
3. The pod's resource usage (memory usage + swap usage) relative to requests (memory requests + swap requests)
```

iholder101 commented:
@haircommander @@jabdoa2 PTAL

jabdoa2 commented:
> @haircommander @@jabdoa2 PTAL

LGTM!

iholder101 commented:
/test pull-kubernetes-node-swap-conformance-fedora-serial
/test pull-kubernetes-node-swap-conformance-ubuntu-serial

iholder101 commented:
> As per the discussions that took part in the KEP update (see discussion [here](https://github.com/kubernetes/enhancements/pull/4701#discussion_r1944859336)), the [eviction order](https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#pod-selection-for-kubelet-eviction) will be defined as follows:
> 
> ```
> The kubelet uses the following parameters to determine the pod eviction order:
> 1. Whether the pod's resource usage (memory usage + swap usage) exceeds requests (memory requests + swap requests)
> 2. [Pod Priority](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/)
> 3. The pod's resource usage (memory usage + swap usage) relative to requests (memory requests + swap requests)
> ```

FYI @yujuhong @dchen1107

iholder101 commented:
> > As per the discussions that took part in the KEP update (see discussion [here](https://github.com/kubernetes/enhancements/pull/4701#discussion_r1944859336)), the [eviction order](https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#pod-selection-for-kubelet-eviction) will be defined as follows:
> > ```
> > The kubelet uses the following parameters to determine the pod eviction order:
> > 1. Whether the pod's resource usage (memory usage + swap usage) exceeds requests (memory requests + swap requests)
> > 2. [Pod Priority](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/)
> > 3. The pod's resource usage (memory usage + swap usage) relative to requests (memory requests + swap requests)
> > ```
> 
> FYI @yujuhong @dchen1107

@yujuhong @dchen1107 this PR is here for two months, and there are 4 days left until feature freeze, and no reviews although the KEP is marked as a high sig-node priority =/

With such a small time-window it would be extremely challenging, to say the least, to address the feedback before freeze.

kannon92 commented:
It does seem like you have gotten some reviews for this.

I still would like to see some test coverage on 

https://github.com/kubernetes/kubernetes/pull/129578#issuecomment-2593075076

Can we address this by adding memoryEvictions to our swap conformance lane and making sure memory based evictions still work with Swap?

iholder101 commented:
> It does seem like you have gotten some reviews for this.
> 
> I still would like to see some test coverage on
> 
> [#129578 (comment)](https://github.com/kubernetes/kubernetes/pull/129578#issuecomment-2593075076)
 
I have received a review comment regarding tests. Your comment links my answer from 2 months ago, in which I raise a suggestion on how to address the concern. There was never a reply to my suggestion.

> Can we address this by adding memoryEvictions to our swap conformance lane and making sure memory based evictions still work with Swap?

Firstly, memory eviction tests are already added to the swap conformance lane as part of this PR. See [this commit](https://github.com/kubernetes/kubernetes/pull/129578/commits/ec277b7c147f6921d85b3ba92edda3de95638e4f).
Secondly, my suggestion from the comment you've linked still holds. Can you share your thoughts regarding it? would it lower your concern?

In any case, this doesn't seem to be the main blocker for this PR.

kannon92 commented:
I think bringing eviction tests into a swap lane but only selectively enabling the memory ones is important. I really would like to see that memory evictions without any swap usage still behave the same way.

iholder101 commented:
> I think bringing eviction tests into a swap lane but only selectively enabling the memory ones is important. I really would like to see that memory evictions without any swap usage still behave the same way.

Right, but do you think it should be part of the swap-conformance lanes, which their main purpose is to execute tests that heavily stress the node? Or, like I suggested, make the "regular" swap lanes (which run the whole test suite on nodes with provisioned swap) run these tests? That's the original question I've asked in the previous comment you've linked. The main rationale is to keep the swap-conformance green, since these kinds of tests are pretty fragile.

I'd be happy to go ahead and introduce a PR for this. However as said before this doesn't seem to be the main blocker for this PR.

kannon92 commented:
If you bring in the memory evictions only, it should be fine. Adding them to the serial jobs should be fine.

kannon92 commented:
We have made a lot of progress in deflating the eviction tests and I haven’t seen any failures for the memory tests.

iholder101 commented:
> If you bring in the memory evictions only, it should be fine. Adding them to the serial jobs should be fine.

PR is ready to review: https://github.com/kubernetes/test-infra/pull/34507.
Let me know what you think.

iholder101 commented:
/test pull-kubernetes-node-swap-conformance-fedora-serial
/test pull-kubernetes-node-swap-conformance-ubuntu-serial

iholder101 commented:
> > If you bring in the memory evictions only, it should be fine. Adding them to the serial jobs should be fine.
> 
> PR is ready to review: [kubernetes/test-infra#34507](https://github.com/kubernetes/test-infra/pull/34507). Let me know what you think.

@kannon92 the PR is now merged, this PR is rebased on top of it, and seems that the swap-conformance lanes are both affected and passing.

Is anything missing from your side for an LGTM?

iholder101 commented:
> I think the code and concept look good to me.
> 
> But I will leave lgtm/approvals for the people that were concerned about this feature. ie google GKE folks.

Thanks

@dchen1107 @yujuhong PTAL

ffromani commented:
/test pull-kubernetes-unit

kannon92 commented:
/retest

kannon92 commented:
btw, https://testgrid.k8s.io/sig-node-kubelet#kubelet-swap-conformance-fedora-serial are flaking.

THese jobs do not contain this code but still worth investigating why evictions on a swap enabled node behave differently.

These jobs are not flaking at all in the main lane (https://testgrid.k8s.io/sig-node-cri-o#ci-crio-cgroupv2-node-e2e-eviction) so swap seems to impact this.

ajaysundark commented on pkg/kubelet/eviction/helpers.go:
```diff
@@ -692,15 +708,42 @@ func exceedMemoryRequests(stats statsFunc) cmpFunc {
+			p1AccessibleSwap, p1Err := swapLimitCalculator.CalcPodSwapLimit(*p1)
+			p2AccessibleSwap, p2Err := swapLimitCalculator.CalcPodSwapLimit(*p2)
+
+			p1MemoryRequest.Add(*resource.NewQuantity(p1AccessibleSwap, resource.BinarySI))
```
This approach of equally treating two unequal resources seems wrong for addressing memory pressure to me.

1. Swap is a less expensive resource, making it easy to imagine larger swap spaces being provisioned than memory. Disk storage/swap availability cannot be compared with memory usage. For example, a user could set up 40GB of swap space in a 4GB physical memory setup. Even when 50% swap is implicitly allocated to a resource, this condition will likely never be satisfied because plenty of swap capacity will still be available.
2. Swap is intended as an overflow mechanism, not primary memory; eviction could trigger before swap utilization even begins, especially in large swap configurations.
3. Swap is significantly slower than memory, and treating both equally for handling memory pressure doesn't sound right. Kubelet's intent for eviction is to handle memory pressure, and if it were to evict a pod that would free up more swap space than memory, it wouldn't help in reducing memory pressure.

The other two criteria (pod priority and relative usage) are reasonable, but they are still affected by this fundamental equivalence problem.

ajaysundark commented:
I think existing eviction behavior: `OrderedBy(exceededMemoryRequests, priority, memory-usage)` is actually more sensible than the proposed approach.

The formula "memory-usage + swap usage > memory requests + swap requests" creates a false equivalence. If swap were to be considered for eviction, better approach would be some form of weighted formula (as suggested by @pacoxu).

Or if we want to simplify without the weights, consider: `OrderedBy(exceededMemoryRequests, priority, memoryUsage + swapUsage)`,  keeping `exceededMemoryRequests` purely based on memory focusing on the actual  memory pressure first.

ajaysundark commented:
Also thinking how does this eviction strategy align with when api is introduced - will the api be at pod or container level? We are rationing swap today at the container level, if API will be at the pod-level, the meaning of 'swap accessibilitiy' for the resources would also change.  Also, KEP suggests considering swap as a separate resource constraint in the future. If swap becomes a first-class resource in Kubernetes with its own requests and limits, it would make sense to consider swap pressure independent from memory pressure.

ajaysundark commented:
As called out in earlier [discussions](https://github.com/kubernetes/enhancements/pull/4701#issuecomment-2657861518), Kubernetes should consider adding I/O pressure as a factor in eviction decisions, which is more relevant when considering swap usage impact on node-stability.

iholder101 commented:
> btw, [testgrid.k8s.io/sig-node-kubelet#kubelet-swap-conformance-fedora-serial](https://testgrid.k8s.io/sig-node-kubelet#kubelet-swap-conformance-fedora-serial) are flaking.
> 
> THese jobs do not contain this code but still worth investigating why evictions on a swap enabled node behave differently.
> 
> These jobs are not flaking at all in the main lane ([testgrid.k8s.io/sig-node-cri-o#ci-crio-cgroupv2-node-e2e-eviction](https://testgrid.k8s.io/sig-node-cri-o#ci-crio-cgroupv2-node-e2e-eviction)) so swap seems to impact this.

This actually makes total sense to me, and was one of the reasons I was concerned from running eviction tests on the swap-conformance lane. The failures are over the node **not** getting into memory pressure, which makes total sense.

I think we need to revert these changes. I've opened a PR here: https://github.com/kubernetes/test-infra/pull/34558.

As explained in the PR:
> The reason is that it causes the lane to become flaky.
Eviction tests are very delicate and fragile. They stress out the node entirely, and depend on many delicate factors like the amount of total/free memory on the node, etc. Even if swap is disabled for Kubernetes workloads, the fact it's enabled on the node might easily cause these tests to become flaky. The swap-conformance stress tests are designed to run on this lane and they are very stable.

iholder101 commented:
> thinking how does this eviction strategy align with when api is introduced - will the api be at pod or container level? We are rationing swap today at the container level, if API will be at the pod-level, the meaning of 'swap accessibilitiy' for the resources would also change. Also, KEP suggests considering swap as a separate resource constraint in the future.

APIs for swap is a complicated discussion for many reasons, like who controls swap limits (would be dangerous to give the pod owner direct control), how to set concrete swap limits, etc. In fact, the KEP explicitly [says](https://github.com/kubernetes/enhancements/blob/1d9ad49058f0114b5b45f9624ad044b9f573c883/keps/sig-node/2400-node-swap/README.md?plain=1#L136):
> This KEP aims to introduce basic swap enablement and leave further extensions to follow-up KEPs.
> ...
> For example, to achieve this goal, this KEP does not introduce any APIs that allow customizing how the feature behaves, but instead only determines whether the feature is enabled or disabled.

I find it very problematic that after sig-node had agreed on this with a large consensus it's now being opened again as if nothing was agreed. It makes sense to me that the bar for swap is very high, and that the burden of proof is on me. However, I think it's essential that once we decide on the scope of the feature we would respect these agreements.

In any case, I'm okay with re-discussing the API as long as we look at it from the perspective of the most minimal API that's needed as a first step for the current KEP to GA, while deferring everything that's not strictly necessary to follow-up KEPs which I'll gladly work on.

> If swap becomes a first-class resource in Kubernetes with its own requests and limits, it would make sense to consider swap pressure independent from memory pressure.

I don't see swap ever becoming a first-class resource, for many reasons. This was already discussed and ruled out in the past.

iholder101 commented:
> As called out in earlier [discussions](https://github.com/kubernetes/enhancements/pull/4701#issuecomment-2657861518), Kubernetes should consider adding I/O pressure as a factor in eviction decisions, which is more relevant when considering swap usage impact on node-stability.

I agree. This sounds like a great topic for a dedicated KEP, as kubelet currently doesn't have any protection mechanism against IO.

I'd go another step forward and say that the eviction manager's code is very old and we may want to frame this follow-up KEP around more general eviction improvements.
